### PR TITLE
ENT-9147 Remove un-notarised transactions upon Double Spend.

### PR DESCRIPTION
**Pull request by colljos**
Two Phase Finality will automatically delete an un-notarised transaction from the `DBTransaction` table when a Double Spend is detected in Initiator `FinalityFlow`.

Furthermore, if the new optional `FinalityFlow` `handleDoubleSpend` constructor parameter is set to `true` (default: `false`), then the Double Spend error (NotaryError.Conflict) will propagate to 2PF peers allowing them to automatically remove the associated un-notarised transaction from their `DBTransaction` table.

If the CorDapp is compiled against 4.11 (e.g. its TPV = 13) then Double Spend handling is enabled by default.

### CODE REVIEW

File: core/src/main/kotlin/net/corda/core/flows/FinalityFlow.kt [hunk 1]
```diff
@@ -111,6 +114,22 @@ class FinalityFlow private constructor(val transaction: SignedTransaction,
...
+     * Notarise the given transaction and broadcast it to all the participants.
+     * Optionally, catch and propagate Double Spend exception (NotaryError.Conflict) to peers.
+     *
+     * @param transaction What to commit.
+     * @param sessions A collection of [FlowSession]s for each non-local participant of the transaction. Sessions to non-participants can
+     * also be provided.
+     * @param statesToRecord Which states to commit to the vault.
+     * @param handleDoubleSpend Whether to catch and propagate Double Spend exception to peers.
+     */
+    constructor(
```
Comment (rick-r3) [inline core/src/main/kotlin/net/corda/core/flows/FinalityFlow.kt]: I think we'll have to do work on the constructors... why not add this (extra param) to the ones above that have the `@JvmOverloads` with the default set to null ?

File: core/src/main/kotlin/net/corda/core/flows/FinalityFlow.kt [hunk 2]
```diff
@@ -202,28 +223,40 @@ class FinalityFlow private constructor(val transaction: SignedTransaction,
...
+                broadcastToOtherParticipants(externalTxParticipants, newPlatformSessions + oldPlatformSessions, stxn)
+            } else if (useTwoPhaseFinality && oldPlatformSessions.isNotEmpty()) {
+                broadcastToOtherParticipants(externalTxParticipants, oldPlatformSessions, stxn)
+            }
+            return stxn
+        }
+        catch (e: NotaryException) {
+            if (e.error is NotaryError.Conflict && useTwoPhaseFinality) {
+                (serviceHub as ServiceHubCoreInternal).removeUnnotarisedTransaction(e.error.txId)
+                val overrideHandleDoubleSpend = handleDoubleSpend ?:
```
Comment (rick-r3) [inline core/src/main/kotlin/net/corda/core/flows/FinalityFlow.kt:240]: I think you need to be careful here, and apply, in addition, the same criteria to sending the error (as a basis) to the condition around `broadcastSignaturesAndFinalise()`.  i.e. `useTwoPhaseFinality && newPlatformSessions.isNotEmpty()`.

i.e. if you are using 2PF and there are new platform sessions

File: core/src/main/kotlin/net/corda/core/flows/FinalityFlow.kt [hunk 3]
```diff
@@ -202,28 +223,40 @@ class FinalityFlow private constructor(val transaction: SignedTransaction,
...
+                    (serviceHub.cordappProvider.getAppContext().cordapp.targetPlatformVersion >= PlatformVersionSwitches.TWO_PHASE_FINALITY)
+                            .apply { logger.warn("Overriding handleDoubleSpend from $handleDoubleSpend to $this") }
```
Comment (rick-r3) [inline core/src/main/kotlin/net/corda/core/flows/FinalityFlow.kt]: Whether we want to warn here or not depends a little bit on the cost of `serviceHub.cordappProvider.getAppContext().cordapp.targetPlatformVersion`.  If it ends up very expensive, we might have to rethink any way.  So I'm going to suggest we don't warn.

I will have a look at the logic in that call as I bet it's pretty dumb and if it is slow we could speed it up.

**Review by rick-r3 — COMMENTED** [review]
Also see slack comment about whether we commit the transaction deletion when we throw an exception right after it.  Need to double check because I don't recall what is supposed to happen there.  We might need to force a checkpoint.

File: core/src/main/kotlin/net/corda/core/flows/FinalityFlow.kt (hunk 2 above)
Comment (colljos) [reply core/src/main/kotlin/net/corda/core/flows/FinalityFlow.kt:240]: `useTwoPhaseFinality` is already checked upon checking the Exception.
I'll add `&& newPlatformSessions.isNotEmpty()` prior to calling broadcast.

File: core/src/main/kotlin/net/corda/core/flows/FinalityFlow.kt (hunk 3 above)
Comment (colljos) [reply core/src/main/kotlin/net/corda/core/flows/FinalityFlow.kt]: The call to `getAppContext()` is expensive (we refrained from using it before) as it performs stackTrace traversal.
Any performance impact will be isolated to failed Double Spends where the default constructor is not set.

Comment (colljos) [reply core/src/main/kotlin/net/corda/core/flows/FinalityFlow.kt]: I'll remove the warning for now.

Comment (colljos) [reply core/src/main/kotlin/net/corda/core/flows/FinalityFlow.kt:253]: > 

Good catch.
A forced checkpoint is required where an Exception is thrown.
I've updated the tests to check the contents of the Db.

**Comment by colljos** [comment]
> Also see slack comment about whether we commit the transaction deletion when we throw an exception right after it. Need to double check because I don't recall what is supposed to happen there. We might need to force a checkpoint.

Good catch.
A forced checkpoint is required where an Exception is thrown.
I've updated the tests to check the contents of the Db.

File: core/src/main/kotlin/net/corda/core/flows/FinalityFlow.kt [hunk 4]
```diff
@@ -202,28 +212,39 @@ class FinalityFlow private constructor(val transaction: SignedTransaction,
...
+            return stxn
+        }
+        catch (e: NotaryException) {
+            if (e.error is NotaryError.Conflict && useTwoPhaseFinality) {
+                (serviceHub as ServiceHubCoreInternal).removeUnnotarisedTransaction(e.error.txId)
+                val overrideHandleDoubleSpend = handleDoubleSpend ?:
+                    (serviceHub.cordappProvider.getAppContext().cordapp.targetPlatformVersion >= PlatformVersionSwitches.TWO_PHASE_FINALITY)
+                if (overrideHandleDoubleSpend && newPlatformSessions.isNotEmpty()) {
+                    broadcastDoubleSpendError(newPlatformSessions, e)
+                }
```
Comment (rick-r3) [inline core/src/main/kotlin/net/corda/core/flows/FinalityFlow.kt]: Re: needing a checkpoint after the removal before throwing, do you need to force a suspend here in an `else` if the `if` branch above that will suspend due to sends isn't taken?

**Comment by rick-r3** [comment]
I see you will now be having "fun" with some of Dan's flow hospital tests that perhaps have different outcomes due to trying to inject failures in the wrong place etc.

**Comment by colljos** [comment]
> I see you will now be having "fun" with some of Dan's flow hospital tests that perhaps have different outcomes due to trying to inject failures in the wrong place etc.

Yes - I make a few tweaks to cater for the new thrown `NotaryException`.
Annoyingly there are 2 tests that fail when the whole file is executed but they pass in isolation.

**Comment by colljos** [comment]
API checker is now complaining upon making the suggested constructor changes:

```
[2023-04-14T08:58:15.875Z] Number of API removals/changes: 2

[2023-04-14T08:58:15.875Z] -  public <init>(net.corda.core.transactions.SignedTransaction, java.util.Collection, net.corda.core.node.StatesToRecord, net.corda.core.utilities.ProgressTracker, int, kotlin.jvm.internal.DefaultConstructorMarker)

[2023-04-14T08:58:15.875Z] -  public <init>(net.corda.core.transactions.SignedTransaction, java.util.Collection, net.corda.core.utilities.ProgressTracker, int, kotlin.jvm.internal.DefaultConstructorMarker)
```

File: core/src/main/kotlin/net/corda/core/flows/FinalityFlow.kt (hunk 4 above)
Comment (colljos) [reply core/src/main/kotlin/net/corda/core/flows/FinalityFlow.kt]: Fixed + added test assertions.

File: core/src/main/kotlin/net/corda/core/flows/FinalityFlow.kt (hunk 1 above)
Comment (colljos) [reply core/src/main/kotlin/net/corda/core/flows/FinalityFlow.kt]: Makes API Checker unhappy: https://ci01.dev.r3.com/blue/organizations/jenkins/Corda-Open-Source%2FCorda-OS-Pull-Request-Code-Checks%2Fcorda/detail/PR-7324/13/pipeline
Should I just update the baseline file?

File: core/src/main/kotlin/net/corda/core/flows/FinalityFlow.kt (hunk 3 above)
Comment (rick-r3) [reply core/src/main/kotlin/net/corda/core/flows/FinalityFlow.kt]: > Any performance impact will be isolated to failed Double Spends 

That is true, so the cost is not so important on the common path.  Looking at that method, it isn't really possible to speed it up, but I think we could offer a similar API in 4.11 that just looks as the sub flow stack which is super cheap.  i.e. which flow were we in when `subFlow(FinalityFlow())` is as good as which CorDapp.  The only real difference is if a flow uses a library from another CorDapp which then calls `FinalityFlow` -  but that would be an okay compromise here.  Anyway, not needed here.

File: .ci/api-current.txt [hunk 5]
```diff
@@ -2542,9 +2542,7 @@ public final class net.corda.core.flows.FinalityFlow extends net.corda.core.flow
   public <init>(net.corda.core.transactions.SignedTransaction, java.util.Collection<? extends net.corda.core.flows.FlowSession>, java.util.Collection<net.corda.core.identity.Party>, net.corda.core.utilities.ProgressTracker)
   public <init>(net.corda.core.transactions.SignedTransaction, java.util.Collection<? extends net.corda.core.flows.FlowSession>, net.corda.core.node.StatesToRecord)
   public <init>(net.corda.core.transactions.SignedTransaction, java.util.Collection<? extends net.corda.core.flows.FlowSession>, net.corda.core.node.StatesToRecord, net.corda.core.utilities.ProgressTracker)
-  public <init>(net.corda.core.transactions.SignedTransaction, java.util.Collection, net.corda.core.node.StatesToRecord, net.corda.core.utilities.ProgressTracker, int, kotlin.jvm.internal.DefaultConstructorMarker)
   public <init>(net.corda.core.transactions.SignedTransaction, java.util.Collection<? extends net.corda.core.flows.FlowSession>, net.corda.core.utilities.ProgressTracker)
-  public <init>(net.corda.core.transactions.SignedTransaction, java.util.Collection, net.corda.core.utilities.ProgressTracker, int, kotlin.jvm.internal.DefaultConstructorMarker)
```
Comment (rick-r3) [inline .ci/api-current.txt:2547]: Yes, I think these are okay because they are just used by the other generated methods, which will have been regenerated.

**Review by rick-r3 — APPROVED** [review]
Merge it!   Let's have a think about the name of the `handleDoubleSpend` parameter though.  It doesn't really convey it changes behaviour of the receiver and folks might think it does more than that.   But don't want to hold up the merge while we think about a name.  Raise a ticket.

**Comment by colljos** [comment]
> Let's have a think about the name of the `handleDoubleSpend` parameter though. It doesn't really convey it changes behaviour of the receiver and folks might think it does more than that. But don't want to hold up the merge while we think about a name. Raise a ticket.

https://r3-cev.atlassian.net/browse/ENT-9823

**Pull request merged**